flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/validate/FunctionCatalog.scala

```diff
@@ -47,13 +52,50 @@ class FunctionCatalog {
     sqlFunctions += sqlFunction
   }
 
+  /** Register multiple sql functions at one time. The functions has the same name. **/
+  def registerSqlFunctions(functions: Seq[SqlFunction]): Unit = {
+    if (functions.nonEmpty) {
+      sqlFunctions --= sqlFunctions.filter(_.getName == functions.head.getName)
+      sqlFunctions ++= functions
+    }
+  }
+
   def getSqlOperatorTable: SqlOperatorTable =
     ChainedSqlOperatorTable.of(
       new BasicOperatorTable(),
       new ListSqlOperatorTable(sqlFunctions)
     )
 
   /**
+    * Lookup table function and create an TableFunctionCall if we find a match.
+    */
+  def lookupTableFunction[T](name: String, children: Seq[Expression]): TableFunctionCall[T] = {
+    val funcClass = functionBuilders
+      .getOrElse(name.toLowerCase, throw ValidationException(s"Undefined function: $name"))
+    funcClass match {
+      // user-defined table function call
+      case tf if classOf[TableFunction[T]].isAssignableFrom(tf) =>
+        Try(UserDefinedFunctionUtils.instantiate(tf.asInstanceOf[Class[TableFunction[T]]])) match {
+          case Success(tableFunction) => {
+            val clazz: Type = tableFunction.getClass.getGenericSuperclass
```

Comment:
As I said earlier this and following lines are very error-prone. We should keep calls to the TypeExtractor to a very minimum. The TypeExtractor is only intended for Java not Scala.